torch/csrc/autograd/python_variable.cpp

```diff
@@ -84,22 +84,41 @@ using torch::autograd::impl::set_pyobj;
 // have a PyObject* associated with it.
 static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
 {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(var.unsafeGetTensorImpl()->pyobj() == nullptr);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(var.unsafeGetTensorImpl()->owns_pyobj() == false);
   PyObject* obj = type->tp_alloc(type, 0);
   if (obj) {
     auto v = (THPVariable*) obj;
-    new (&v->cdata) Variable(std::move(var));
-    set_pyobj(v->cdata, obj);
+    // TODO: named constructor to avoid default initialization
+    new (&v->cdata) MaybeOwned<Variable>();
+    v->cdata = MaybeOwned<Variable>::owned(std::move(var));
+    // Can't use var as it's been moved out of
+    set_pyobj(THPVariable_Unpack(v), obj);
   }
   return obj;
 }
 
+// TODO: Make this take Variable by const reference
 PyObject * THPVariable_Wrap(Variable var)
 {
   if (!var.defined()) {
     Py_RETURN_NONE;
   }
 
   if (auto obj = pyobj(var)) {
+    if (var.unsafeGetTensorImpl()->owns_pyobj()) {
+      TORCH_INTERNAL_ASSERT(Py_REFCNT(obj) == 1);
```

Comment:
What if there are Python weak references to the Python tensor? Can't those increase the refcount (by becoming strong references) while it's a zombie?

Reply:
Yes, I think so. Ugh, actually I have no idea how to solve this, the original scheme was predicated on being able to intercept all refcount increases from refcount 1.

One possibility is to just leave this as is, and say that the C++ implementation is allowed to die in the meantime until you call some special function that revives the function. Not very satisfactory...

Reply:
One possibility is we can provide our own function for dereferencing Python weak references that handles ownership flipping correctly.

Another possibility is to "fake" resurrect the C++ object in this case. Ordinarily, when we deallocate a Tensor, we just release the Python object immediately. We could have decref method instead test if the Python object is still alive or dead. If it is still alive, we do a shallow copy of Tensor into a new Tensor, stuff *that* into the Python object (owning reference), but let the current Tensor continue dying. This will break C++ side weak references (which will go dead even though it's truly dead) but it should at least not result in any crashes. WDYT.